Makefile

```diff
@@ -12,72 +12,62 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-SHELL = /bin/bash
-
-GO_VERSION = 1.12.5
+GO_VERSION = 1.13.8
 
 export CGO_ENABLED := 0
 
-export E2E_SSH_PUBKEY ?= $(shell cat ~/.ssh/id_rsa.pub)
+export E2E_SSH_PUBKEY ?= $(shell test -f ~/.ssh/id_rsa.pub && cat ~/.ssh/id_rsa.pub)
 
-export GIT_TAG ?= $(shell git tag --points-at HEAD)
+export DOCKER_TAG ?= $(shell git tag --points-at HEAD)
 
 REGISTRY ?= docker.io
 REGISTRY_NAMESPACE ?= kubermatic
 
+LDFLAGS ?= -ldflags '-s -w'
+
 IMAGE_TAG = \
 		$(shell echo $$(git rev-parse HEAD && if [[ -n $$(git status --porcelain) ]]; then echo '-dirty'; fi)|tr -d ' ')
 IMAGE_NAME = $(REGISTRY)/$(REGISTRY_NAMESPACE)/machine-controller:$(IMAGE_TAG)
 
+OS = centos coreos ubuntu sles rhel
+USERDATA_BIN = $(patsubst %, machine-controller-userdata-%, $(OS))
+
+.PHONY: all
+all: machine-controller webhook $(USERDATA_BIN)
+
+machine-controller-userdata-%: cmd/userdata/% $(shell find cmd/userdata/$* pkg -name '*.go')
+	go build -v \
+		$(LDFLAGS) \
+		-o $@ \
+		github.com/kubermatic/machine-controller/cmd/userdata/$*
+
+%: cmd/% $(shell find cmd/$* pkg -name '*.go')
+	go build -v \
+		$(LDFLAGS) \
+		-o $@ \
+		github.com/kubermatic/machine-controller/cmd/$*
 
-vendor: Gopkg.lock Gopkg.toml
-	dep ensure -vendor-only
+.PHONY: clean
+clean:
+	rm -f machine-controller \
+		webhook \
+		$(USERDATA_BIN)
 
+.PHONY: machine-controller-docker
 machine-controller-docker:
 	@docker run --rm \
 		-v $$PWD:/go/src/github.com/kubermatic/machine-controller \
 		-v $$PWD/.buildcache:/cache \
 		-e GOCACHE=/cache \
 		-w /go/src/github.com/kubermatic/machine-controller \
 		golang:$(GO_VERSION) \
-			make machine-controller
-
-.PHONY: machine-controller
-machine-controller: $(shell find cmd pkg -name '*.go') vendor
-	go build -v \
-		-ldflags '-s -w' \
-		-o machine-controller \
-		github.com/kubermatic/machine-controller/cmd/controller
-	go build -v \
-		-ldflags '-s -w' \
-		-o machine-controller-userdata-centos \
-		github.com/kubermatic/machine-controller/cmd/userdata/centos
-	go build -v \
-		-ldflags '-s -w' \
-		-o machine-controller-userdata-coreos \
-		github.com/kubermatic/machine-controller/cmd/userdata/coreos
-	go build -v \
-		-ldflags '-s -w' \
-		-o machine-controller-userdata-ubuntu \
-		github.com/kubermatic/machine-controller/cmd/userdata/ubuntu
-	go build -v \
-		-ldflags '-s -w' \
-		-o machine-controller-userdata-sles \
-		github.com/kubermatic/machine-controller/cmd/userdata/sles
-	go build -v \
-		-ldflags '-s -w' \
-		-o machine-controller-userdata-rhel \
-		github.com/kubermatic/machine-controller/cmd/userdata/rhel
-
-webhook: $(shell find cmd pkg -name '*.go') vendor
-	go build -v \
-		-ldflags '-s -w' \
-		-o webhook \
-		github.com/kubermatic/machine-controller/cmd/webhook
+			make machine-controller CGO_ENABLED="$(CGO_ENABLED)" MODFLAG="$(MODFLAG)" E2E_SSH_PUBKEY=
```

Comment:
I'm not a makefile expert, but I'm wondering what is the reason for overriding `E2E_SSH_PUBKEY` with empty value? Or is something else happening here?

Reply:
good point, let me explain that because it isn't obvious. We are defaulting the [E2E_SSH_PUBKEY](https://github.com/kubermatic/machine-controller/pull/680/files#diff-b67911656ef5d18c4ae36cb6741b7965L21) variable by running this shell command `cat ~/.ssh/id_rsa.pub`, unless a value is provided, and it can be empty. In targets `machine-controller-docker` and  `test-unit` we're invoking make inside a container and I noted that we were not forwarding the values that could have been provided in input for `CGO_ENABLED` and `E2E_SSH_PUBKEY`. This was a bug and I fixed it. Note that the `E2E_SSH_PUBKEY` is not required for both targets, the only reason why I specified an empty value is to avoid evaluating the variable and polluting the output with:
```
cat: /root/.ssh/id_rsa.pub: No such file or directory
```
That said I later added [this](https://github.com/kubermatic/machine-controller/pull/680/files#diff-b67911656ef5d18c4ae36cb6741b7965R19) test, that makes this fix useless, so I'll remove it.
I also spotted on the same line that the target that should be run is now `all` and not `machine-controller`.

Reply:
Makes sense, thank you for such a detailed explanation!